R/mle_finder.R

```diff
@@ -22,3 +22,52 @@ find_mle_linear_pseudo_inv = function(design, outcome) {
 
     as.vector(backsolve(upper, backsolve(upper, b, transpose = TRUE)))
 }
+
+find_mle_logit_bfgs = function(design, outcome) {
+
+    n_pred = ncol(design)
+    coeff_init = rep(1, n_pred)
+
+    op_result = optim(coeff_init,
+                      log_likelihood_logit,
+                      log_likelihood_logit_gradient,
+                      design = design,
+                      outcome = outcome,
+                      method = "BFGS",
+                      control = list(fnscale = -1))
+
+    return(op_result$par)
+}
+
+find_mle_logit_newton = function(design, outcome, max_iter = 10000, init = rep(0, ncol(design))) {
+
+    coeff = init
+    iter = 0
+    abs_tol = 1e-6
+    rel_tol = 1e-6
+
+    while (iter < max_iter) {
+        pre_loglikelihood = log_likelihood_logit(design, outcome, coeff)
+
+        hessian = log_likelihood_logit_hessian(design, outcome, coeff)
+        gradient = log_likelihood_logit_gradient(design, outcome, coeff)
+        coeff = coeff - solve(hessian) %*% gradient
```

Comment:
```suggestion
        coeff = coeff - solve(hessian,  gradient)
```

-2

Recall from HW2 that you shouldnt invert matrices. You could also use your cholesky function you wrote previously.